src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderingHelper.cs

```diff
@@ -166,9 +168,82 @@ public static IProjectTree GetLastChild(IProjectTree projectTree)
         /// </summary>
         public static IProjectTree GetFirstChild(IProjectTree projectTree)
         {
+            Requires.NotNull(projectTree, nameof(projectTree));
+
             return GetChildren(projectTree).FirstOrDefault();
         }
 
+        /// <summary>
+        /// Get the difference to see what was added.
+        /// We do a sanity check to make sure they have a valid display order.
+        /// We also order the added nodes by their display order.
+        /// </summary>
+        public static ReadOnlyCollection<IProjectTree> GetAddedNodes(IProjectTree previousTarget, IProjectTree currentTarget)
+        {
+            Requires.NotNull(previousTarget, nameof(previousTarget));
+            Requires.NotNull(currentTarget, nameof(currentTarget));
+
+            return currentTarget.Children
+                .Where(updatedChild => IsAddedNode(previousTarget, updatedChild))
+                .OrderBy(GetDisplayOrder)
+                .ToList()
+                .AsReadOnly();
```

Comment:
This is very inefficient allocation-wise, it creates a `ReadOnlyCollection<T>` that wraps a `List<T>` that wraps an `T[]`, plus the other 3 allocations for the Linq query. Can we just put this straight into an ImmutableList<T> via `ToImmutableList()/ToImmutableArray()` to reduce some of that?

Reply:
Yes, `ToImmutableArray` should be fine.

Reply:
Done, also changed a similar method to do the same.